### PR TITLE
Add entr fallback

### PR DESCRIPTION
**Pull request by JimMadge**
Add an alternative to using `entr` to monitor allowlist changes.

Activated by defining the `ENTR_FALLBACK` environment variable.

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -41,5 +45,22 @@ if ! nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEX
...
+if [ -n "$ENTR_FALLBACK" ]; then
+    echo "$(timestamp) Using fallback file monitoring"
+    # Run allowlist configuration now
+    nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEXUS_HOST" --nexus-port "$NEXUS_PORT" update-allowlists --packages "$NEXUS_PACKAGES" --pypi-package-file "$PYPI_ALLOWLIST" --cran-package-file "$CRAN_ALLOWLIST"
+    # Periodically check for modification of allowlist files and run configuration again when they are
+    hash=$(hashes)
+    while true; do
+        new_hash=$(hashes)
+        if [ "$hash" != "$new_hash" ]; then
+            nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEXUS_HOST" --nexus-port "$NEXUS_PORT" update-allowlists --packages "$NEXUS_PACKAGES" --pypi-package-file "$PYPI_ALLOWLIST" --cran-package-file "$CRAN_ALLOWLIST"
```
Comment (jemrobinson) [inline entrypoint.sh:57]: We should wrap this stuff in a function rather than replicating it.

File: entrypoint.sh [hunk 2]
```diff
@@ -41,5 +45,22 @@ if ! nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEX
...
-find "$ALLOWLIST_DIR"/*.allowlist | entr -n nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEXUS_HOST" --nexus-port "$NEXUS_PORT" update-allowlists --packages "$NEXUS_PACKAGES" --pypi-package-file "$PYPI_ALLOWLIST" --cran-package-file "$CRAN_ALLOWLIST"
+if [ -n "$ENTR_FALLBACK" ]; then
+    echo "$(timestamp) Using fallback file monitoring"
+    # Run allowlist configuration now
+    nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEXUS_HOST" --nexus-port "$NEXUS_PORT" update-allowlists --packages "$NEXUS_PACKAGES" --pypi-package-file "$PYPI_ALLOWLIST" --cran-package-file "$CRAN_ALLOWLIST"
+    # Periodically check for modification of allowlist files and run configuration again when they are
+    hash=$(hashes)
+    while true; do
+        new_hash=$(hashes)
+        if [ "$hash" != "$new_hash" ]; then
```
Comment (jemrobinson) [inline entrypoint.sh:56]: Not sure how well `hashes` will compare here as the string will include new lines and look something like:

```
1504b4d2daae6afcaf5d9a396e580ba2  /allowlists/cran.allowlist
fdf8764e4d5f0583abcb5753df8204f2  /allowlists/pypi.allowlist
```

Have you tested that this works?

**Review by jemrobinson — APPROVED** [review]
Some minor QOL changes

File: entrypoint.sh (hunk 1 above)
Comment (JimMadge) [reply entrypoint.sh:57]: Yes, I don't like it.

It seems that the argument for `entr` needs to be a file though. So, at least for the `entr` command, you cannot substitute the command or use a function defined in the script.

I'm not sure whether partially duplicating the command is much better than always duplicating it 🤔.

Comment (jemrobinson) [reply entrypoint.sh:57]: Could we not use a function like this?

```bash
update_allowlists() {
    nexus-allowlist --admin-password "$NEXUS_ADMIN_PASSWORD" --nexus-host "$NEXUS_HOST" --nexus-port "$NEXUS_PORT" update-allowlists --packages "$NEXUS_PACKAGES" --pypi-package-file "$PYPI_ALLOWLIST" --cran-package-file "$CRAN_ALLOWLIST"
}
```

File: entrypoint.sh (hunk 2 above)
Comment (JimMadge) [reply entrypoint.sh:56]: Yep :+1:, it works for either file.
I imagine it is just comparing strings which happen to have `\n` in them.

File: entrypoint.sh (hunk 1 above)
Comment (JimMadge) [reply entrypoint.sh:60]: [It seems not](https://github.com/eradman/entr/issues/6).

I'm not sure moving the command to a new script is much better as you would need to pass some environment variables as arguments, which may make what is going on less clear.

Comment (jemrobinson) [reply entrypoint.sh:60]: Fair enough. Let's deal with the duplication then.

**Pull request merged**